include/inexor/vulkan-renderer/world/collision.hpp

```diff
@@ -0,0 +1,250 @@
+    cube_corner m_nearest_corner;
+    cube_edge m_nearest_edge;
+
+public:
+    /// @brief Calculate point of intersection, selected face,
+    /// nearest corner on that face, and nearest edge on that face.
+    /// @param cube The cube to check for collision.
+    /// @param ray_pos The start point of the ray.
+    /// @param ray_dir The direction of the ray.
+    RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+        // This lambda adjusts the center points on a cube's face to the size of the octree,
+        // so collision works with cubes of any size. This does not yet account for rotations!
+        const auto adjust_coordinates = [=](const glm::vec3 pos) {
+            // TODO: Take rotation of the cube into account.
+            return m_cube.center() + pos * (m_cube.size() / 2);
+        };
+
+        const std::array<glm::vec3, 6> cube_directions{
+            glm::vec3(-1.0f, 0.0f, 0.0f), // left
+            glm::vec3(1.0f, 0.0f, 0.0f),  // right
+            glm::vec3(0.0f, -1.0f, 0.0f), // front
+            glm::vec3(0.0f, 1.0f, 0.0f),  // back
+            glm::vec3(0.0f, 0.0f, 1.0f),  // top
+            glm::vec3(0.0f, 0.0f, -1.0f)  // bottom
+        };
+
+        // The coordinates of the center of every face of the cube.
+        const std::array<cube_face, 6> cube_face_centers{cube_face{"left", adjust_coordinates(cube_directions[0])},
+                                                         cube_face{"right", adjust_coordinates(cube_directions[1])},
+                                                         cube_face{"front", adjust_coordinates(cube_directions[2])},
+                                                         cube_face{"back", adjust_coordinates(cube_directions[3])},
+                                                         cube_face{"top", adjust_coordinates(cube_directions[4])},
+                                                         cube_face{"bottom", adjust_coordinates(cube_directions[5])}};
+
+        // The coordinates of every corner of the cube.
+        std::array<cube_corner, 8> cube_corners = {
+            cube_corner{"left back top", adjust_coordinates({-1.0f, 1.0f, 1.0f})},
+            cube_corner{"right back top", adjust_coordinates({1.0f, 1.0f, 1.0f})},
+            cube_corner{"left back botom", adjust_coordinates({-1.0f, 1.0f, -1.0f})},
+            cube_corner{"right back bottom", adjust_coordinates({1.0f, 1.0f, -1.0f})},
+            cube_corner{"left front top", adjust_coordinates({-1.0f, -1.0f, 1.0f})},
+            cube_corner{"right front top", adjust_coordinates({1.0f, -1.0f, 1.0f})},
+            cube_corner{"left front bottom", adjust_coordinates({-1.0f, -1.0f, -1.0f})},
+            cube_corner{"right front bottom", adjust_coordinates({1.0f, -1.0f, -1.0f})}};
+
+        using corner_on_face_index = std::array<std::size_t, 4>;
+
+        // These indices specify which 4 points describe the corners on the given face.
```

Comment:
In which order? clock wise or counter clock wise?

Reply:
Excellent point! I forgot about this! It should be consistent with the decisions made in octree. @IceflowRE @movabo

Reply:
https://inexor-vulkan-renderer.readthedocs.io/en/latest/development/reference/octree-file-format.html#octree

Reply:
* [ ] TODO: Check if order of corners matches octree code.

Reply:
Done.